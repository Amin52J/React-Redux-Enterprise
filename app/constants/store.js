import {routerReducer} from 'react-router-redux';
import {createStore, applyMiddleware} from 'redux';
import {createEpicMiddleware} from 'redux-observable';
import {persistCombineReducers} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import hocReducer from '@hoc/cleanOnUnmount/reducer';
import * as reducers from '@reducers';
import epics from '@epics';
import {ajax} from 'rxjs/observable/dom/ajax';
import {of} from 'rxjs/observable/of';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/takeUntil';
import {routerMiddleware} from 'react-router-redux';
import createHistory from 'history/createBrowserHistory';

export const history = createHistory();

const persistConfig = {
  key: 'root',
  storage
};
const epicMiddleware = createEpicMiddleware(epics, {
  dependencies: {
    getJSON: ajax.getJSON,
    of
  }
});
const combinedReducers = persistCombineReducers(persistConfig, {
  ...reducers,
  routing: routerReducer
});
const reducer = hocReducer(combinedReducers);
const store = createStore(reducer, applyMiddleware(epicMiddleware, routerMiddleware(history)));

export default store;
