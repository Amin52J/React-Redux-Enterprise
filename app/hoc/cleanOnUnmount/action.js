import * as types from '@constants/actionTypes';

/**
 * hoc/cleanOnUnmount/action.js : RESET action factory
 * @returns {Object} the reset action object
 **/
export function reset() {
  return {
    type: types.RESET
  };
}
