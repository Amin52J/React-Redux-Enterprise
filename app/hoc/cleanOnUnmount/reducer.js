import * as types from '@constants/actionTypes';

/**
 * hoc/cleanOnUnmount/reducer.js : the cleanOnUnmount HOC reducer factory
 * @param {Function} combinedReducers the current reducers factory of the component
 * @returns {Function} combination of the current reducer factory and the new (with reset functionality) reducer factory
 **/
export default combinedReducers => (state, action) => {
  switch (action.type) {
    case types.RESET:
      state = undefined; // eslint-disable-line no-undefined
      break;
    default:
      // do nothing
      break;
  }
  return combinedReducers(state, action);
};
