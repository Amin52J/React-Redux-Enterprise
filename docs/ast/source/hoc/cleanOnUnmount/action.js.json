{
  "type": "File",
  "start": 0,
  "end": 226,
  "loc": {
    "start": {
      "line": 1,
      "column": 0
    },
    "end": {
      "line": 12,
      "column": 0
    }
  },
  "program": {
    "type": "Program",
    "start": 0,
    "end": 226,
    "loc": {
      "start": {
        "line": 1,
        "column": 0
      },
      "end": {
        "line": 12,
        "column": 0
      }
    },
    "sourceType": "module",
    "body": [
      {
        "type": "ImportDeclaration",
        "start": 0,
        "end": 48,
        "loc": {
          "start": {
            "line": 1,
            "column": 0
          },
          "end": {
            "line": 1,
            "column": 48
          }
        },
        "specifiers": [
          {
            "type": "ImportNamespaceSpecifier",
            "start": 7,
            "end": 17,
            "loc": {
              "start": {
                "line": 1,
                "column": 7
              },
              "end": {
                "line": 1,
                "column": 17
              }
            },
            "local": {
              "type": "Identifier",
              "start": 12,
              "end": 17,
              "loc": {
                "start": {
                  "line": 1,
                  "column": 12
                },
                "end": {
                  "line": 1,
                  "column": 17
                },
                "identifierName": "types"
              },
              "name": "types"
            }
          }
        ],
        "source": {
          "type": "StringLiteral",
          "start": 23,
          "end": 47,
          "loc": {
            "start": {
              "line": 1,
              "column": 23
            },
            "end": {
              "line": 1,
              "column": 47
            }
          },
          "extra": {
            "rawValue": "@constants/actionTypes",
            "raw": "'@constants/actionTypes'"
          },
          "value": "@constants/actionTypes"
        },
        "trailingComments": [
          {
            "type": "CommentBlock",
            "value": "*\n * hoc/cleanOnUnmount/action.js : RESET action factory\n * @returns {Object} the reset action object\n *",
            "start": 50,
            "end": 158,
            "loc": {
              "start": {
                "line": 3,
                "column": 0
              },
              "end": {
                "line": 6,
                "column": 4
              }
            }
          }
        ]
      },
      {
        "type": "ExportNamedDeclaration",
        "start": 159,
        "end": 225,
        "loc": {
          "start": {
            "line": 7,
            "column": 0
          },
          "end": {
            "line": 11,
            "column": 1
          }
        },
        "specifiers": [],
        "source": null,
        "declaration": {
          "type": "FunctionDeclaration",
          "start": 166,
          "end": 225,
          "loc": {
            "start": {
              "line": 7,
              "column": 7
            },
            "end": {
              "line": 11,
              "column": 1
            }
          },
          "id": {
            "type": "Identifier",
            "start": 175,
            "end": 180,
            "loc": {
              "start": {
                "line": 7,
                "column": 16
              },
              "end": {
                "line": 7,
                "column": 21
              },
              "identifierName": "reset"
            },
            "name": "reset",
            "leadingComments": null
          },
          "generator": false,
          "expression": false,
          "async": false,
          "params": [],
          "body": {
            "type": "BlockStatement",
            "start": 183,
            "end": 225,
            "loc": {
              "start": {
                "line": 7,
                "column": 24
              },
              "end": {
                "line": 11,
                "column": 1
              }
            },
            "body": [
              {
                "type": "ReturnStatement",
                "start": 187,
                "end": 223,
                "loc": {
                  "start": {
                    "line": 8,
                    "column": 2
                  },
                  "end": {
                    "line": 10,
                    "column": 4
                  }
                },
                "argument": {
                  "type": "ObjectExpression",
                  "start": 194,
                  "end": 222,
                  "loc": {
                    "start": {
                      "line": 8,
                      "column": 9
                    },
                    "end": {
                      "line": 10,
                      "column": 3
                    }
                  },
                  "properties": [
                    {
                      "type": "ObjectProperty",
                      "start": 200,
                      "end": 217,
                      "loc": {
                        "start": {
                          "line": 9,
                          "column": 4
                        },
                        "end": {
                          "line": 9,
                          "column": 21
                        }
                      },
                      "method": false,
                      "shorthand": false,
                      "computed": false,
                      "key": {
                        "type": "Identifier",
                        "start": 200,
                        "end": 204,
                        "loc": {
                          "start": {
                            "line": 9,
                            "column": 4
                          },
                          "end": {
                            "line": 9,
                            "column": 8
                          },
                          "identifierName": "type"
                        },
                        "name": "type"
                      },
                      "value": {
                        "type": "MemberExpression",
                        "start": 206,
                        "end": 217,
                        "loc": {
                          "start": {
                            "line": 9,
                            "column": 10
                          },
                          "end": {
                            "line": 9,
                            "column": 21
                          }
                        },
                        "object": {
                          "type": "Identifier",
                          "start": 206,
                          "end": 211,
                          "loc": {
                            "start": {
                              "line": 9,
                              "column": 10
                            },
                            "end": {
                              "line": 9,
                              "column": 15
                            },
                            "identifierName": "types"
                          },
                          "name": "types"
                        },
                        "property": {
                          "type": "Identifier",
                          "start": 212,
                          "end": 217,
                          "loc": {
                            "start": {
                              "line": 9,
                              "column": 16
                            },
                            "end": {
                              "line": 9,
                              "column": 21
                            },
                            "identifierName": "RESET"
                          },
                          "name": "RESET"
                        },
                        "computed": false
                      }
                    }
                  ]
                }
              }
            ],
            "directives": []
          },
          "leadingComments": [
            {
              "type": "CommentBlock",
              "value": "*\n * hoc/cleanOnUnmount/action.js : RESET action factory\n * @returns {Object} the reset action object\n *",
              "start": 50,
              "end": 158,
              "loc": {
                "start": {
                  "line": 3,
                  "column": 0
                },
                "end": {
                  "line": 6,
                  "column": 4
                }
              }
            }
          ],
          "trailingComments": []
        },
        "leadingComments": [
          {
            "type": "CommentBlock",
            "value": "*\n * hoc/cleanOnUnmount/action.js : RESET action factory\n * @returns {Object} the reset action object\n *",
            "start": 50,
            "end": 158,
            "loc": {
              "start": {
                "line": 3,
                "column": 0
              },
              "end": {
                "line": 6,
                "column": 4
              }
            }
          }
        ]
      }
    ],
    "directives": []
  },
  "comments": [
    {
      "type": "CommentBlock",
      "value": "*\n * hoc/cleanOnUnmount/action.js : RESET action factory\n * @returns {Object} the reset action object\n *",
      "start": 50,
      "end": 158,
      "loc": {
        "start": {
          "line": 3,
          "column": 0
        },
        "end": {
          "line": 6,
          "column": 4
        }
      }
    }
  ],
  "tokens": [
    {
      "type": {
        "label": "import",
        "keyword": "import",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "value": "import",
      "start": 0,
      "end": 6,
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 1,
          "column": 6
        }
      }
    },
    {
      "type": {
        "label": "*",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": 10,
        "updateContext": null
      },
      "value": "*",
      "start": 7,
      "end": 8,
      "loc": {
        "start": {
          "line": 1,
          "column": 7
        },
        "end": {
          "line": 1,
          "column": 8
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "as",
      "start": 9,
      "end": 11,
      "loc": {
        "start": {
          "line": 1,
          "column": 9
        },
        "end": {
          "line": 1,
          "column": 11
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "types",
      "start": 12,
      "end": 17,
      "loc": {
        "start": {
          "line": 1,
          "column": 12
        },
        "end": {
          "line": 1,
          "column": 17
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "from",
      "start": 18,
      "end": 22,
      "loc": {
        "start": {
          "line": 1,
          "column": 18
        },
        "end": {
          "line": 1,
          "column": 22
        }
      }
    },
    {
      "type": {
        "label": "string",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "value": "@constants/actionTypes",
      "start": 23,
      "end": 47,
      "loc": {
        "start": {
          "line": 1,
          "column": 23
        },
        "end": {
          "line": 1,
          "column": 47
        }
      }
    },
    {
      "type": {
        "label": ";",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 47,
      "end": 48,
      "loc": {
        "start": {
          "line": 1,
          "column": 47
        },
        "end": {
          "line": 1,
          "column": 48
        }
      }
    },
    {
      "type": "CommentBlock",
      "value": "*\n * hoc/cleanOnUnmount/action.js : RESET action factory\n * @returns {Object} the reset action object\n *",
      "start": 50,
      "end": 158,
      "loc": {
        "start": {
          "line": 3,
          "column": 0
        },
        "end": {
          "line": 6,
          "column": 4
        }
      }
    },
    {
      "type": {
        "label": "export",
        "keyword": "export",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "value": "export",
      "start": 159,
      "end": 165,
      "loc": {
        "start": {
          "line": 7,
          "column": 0
        },
        "end": {
          "line": 7,
          "column": 6
        }
      }
    },
    {
      "type": {
        "label": "function",
        "keyword": "function",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "function",
      "start": 166,
      "end": 174,
      "loc": {
        "start": {
          "line": 7,
          "column": 7
        },
        "end": {
          "line": 7,
          "column": 15
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "reset",
      "start": 175,
      "end": 180,
      "loc": {
        "start": {
          "line": 7,
          "column": 16
        },
        "end": {
          "line": 7,
          "column": 21
        }
      }
    },
    {
      "type": {
        "label": "(",
        "beforeExpr": true,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 180,
      "end": 181,
      "loc": {
        "start": {
          "line": 7,
          "column": 21
        },
        "end": {
          "line": 7,
          "column": 22
        }
      }
    },
    {
      "type": {
        "label": ")",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 181,
      "end": 182,
      "loc": {
        "start": {
          "line": 7,
          "column": 22
        },
        "end": {
          "line": 7,
          "column": 23
        }
      }
    },
    {
      "type": {
        "label": "{",
        "beforeExpr": true,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 183,
      "end": 184,
      "loc": {
        "start": {
          "line": 7,
          "column": 24
        },
        "end": {
          "line": 7,
          "column": 25
        }
      }
    },
    {
      "type": {
        "label": "return",
        "keyword": "return",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "value": "return",
      "start": 187,
      "end": 193,
      "loc": {
        "start": {
          "line": 8,
          "column": 2
        },
        "end": {
          "line": 8,
          "column": 8
        }
      }
    },
    {
      "type": {
        "label": "{",
        "beforeExpr": true,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 194,
      "end": 195,
      "loc": {
        "start": {
          "line": 8,
          "column": 9
        },
        "end": {
          "line": 8,
          "column": 10
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "type",
      "start": 200,
      "end": 204,
      "loc": {
        "start": {
          "line": 9,
          "column": 4
        },
        "end": {
          "line": 9,
          "column": 8
        }
      }
    },
    {
      "type": {
        "label": ":",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 204,
      "end": 205,
      "loc": {
        "start": {
          "line": 9,
          "column": 8
        },
        "end": {
          "line": 9,
          "column": 9
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "types",
      "start": 206,
      "end": 211,
      "loc": {
        "start": {
          "line": 9,
          "column": 10
        },
        "end": {
          "line": 9,
          "column": 15
        }
      }
    },
    {
      "type": {
        "label": ".",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 211,
      "end": 212,
      "loc": {
        "start": {
          "line": 9,
          "column": 15
        },
        "end": {
          "line": 9,
          "column": 16
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "RESET",
      "start": 212,
      "end": 217,
      "loc": {
        "start": {
          "line": 9,
          "column": 16
        },
        "end": {
          "line": 9,
          "column": 21
        }
      }
    },
    {
      "type": {
        "label": ",",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 217,
      "end": 218,
      "loc": {
        "start": {
          "line": 9,
          "column": 21
        },
        "end": {
          "line": 9,
          "column": 22
        }
      }
    },
    {
      "type": {
        "label": "}",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 221,
      "end": 222,
      "loc": {
        "start": {
          "line": 10,
          "column": 2
        },
        "end": {
          "line": 10,
          "column": 3
        }
      }
    },
    {
      "type": {
        "label": ";",
        "beforeExpr": true,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 222,
      "end": 223,
      "loc": {
        "start": {
          "line": 10,
          "column": 3
        },
        "end": {
          "line": 10,
          "column": 4
        }
      }
    },
    {
      "type": {
        "label": "}",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "start": 224,
      "end": 225,
      "loc": {
        "start": {
          "line": 11,
          "column": 0
        },
        "end": {
          "line": 11,
          "column": 1
        }
      }
    },
    {
      "type": {
        "label": "eof",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 226,
      "end": 226,
      "loc": {
        "start": {
          "line": 12,
          "column": 0
        },
        "end": {
          "line": 12,
          "column": 0
        }
      }
    }
  ]
}