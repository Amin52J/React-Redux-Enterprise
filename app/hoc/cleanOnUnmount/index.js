// Inheritance Inversion HOC

import * as hocActions from '@hoc/cleanOnUnmount/action';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';

function mapStateToProps(state) {
  return state;
}

function reduxConnect(WrappedComponent, actions) {
  return connect(mapStateToProps, actions)(WrappedComponent);
}


export default function (WrappedComponent, actions) {
  const Component = reduxConnect(WrappedComponent, actions);

  class Enhancer extends Component {
    componentWillUnmount() {
      this.props.reset();
      if (super.componentWillUnmount) {
        super.componentWillUnmount();
      }
    }
  }

  return withRouter(reduxConnect(Enhancer, hocActions));
}
