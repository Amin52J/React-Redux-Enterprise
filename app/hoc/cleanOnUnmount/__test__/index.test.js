import React from 'react';
import {mount} from 'enzyme';
import cleanOnUnmount from '../index';
import renderer from 'react-test-renderer';
import store, {history} from '../../../constants/store';
import {ConnectedRouter} from 'react-router-redux';
import {Provider} from 'react-redux';

class Test extends React.Component {
  render() {
    return (
      <div className="clean-on-unmount">cleanOnUnmount</div>
    );
  }
}

const AttachedTest = cleanOnUnmount(Test);

describe('cleanOnUnmount HOC - HOC', () => {
  it('renders with no problem', () => {
    const component = mount(
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <AttachedTest/>
        </ConnectedRouter>
      </Provider>
    );
    const tree = renderer.create(component).toJSON();

    expect(component.find('div').first().hasClass('clean-on-unmount')).toEqual(true);
    expect(tree).toMatchSnapshot();
  });
});
