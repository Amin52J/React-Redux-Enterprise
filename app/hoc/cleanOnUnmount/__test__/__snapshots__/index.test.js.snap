// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanOnUnmount HOC - HOC renders with no problem 1`] = `
<div
  className="clean-on-unmount"
>
  cleanOnUnmount
</div>
`;
